### DISCUSSION TITLE
Handle assets that are too big to fit in memory

### DISCUSSION THREAD

**Discussion by AlejandroUPC** [dagster-io/dagster#17024] [Q&A] [answered]
Let´s say I have three assets representing a basic ETL process: one extracts the data (a queue), some transformation and finally materializing it.
The first step is not an issue (regarding memory)  but afte the transformations and write it becomes a problem.

I cannot control the volume of data that comes from the first step, extraction, but I have to ingest it all in one run, even if it takes longer.

If my pseudo-code is something like this:
```python
from dagster import asset

@asset
def  read_from_stream():
    ....

@asset 
def process_messages():
    ...

@asset 
def write_to_db():
     ....
```

Is there a way I can have this pipeline to process the message in let's say batches of 100 from the first asset? 
I am having a hard time wrapping my head about how to do this with Dagster, I could always reduce the frequency of the round and have a fix set of messages to be parsed (this would not ensure that all of the messages are consumed per run).

Thanks in advance

**Comment by PedramNavid** [comment, marked as answer]
Hi @AlejandroUPC 

How you solve this will depend on where you are persisting data, but nothing is stopping you from batching within an asset itself. I would suggest a slight change to your assets, as they appear to be written as tasks right now (e.g. read from stream, process message, write to db). An asset should reflect the thing you care about, which in this case sounds like a processed message.

Using your example, you might consider having one asset to manage reading, processing, and writing. 

```
from dagster import asset

@asset(deps=[some_source])
def processed_messages(stream: StreamResource, storage: StorageResource):
    for batch in stream.get_batch(): 
        processed = process_batch(batch)
        storage.save_files(processed)
```

You will want to create resources, here's an example of a resource for the StreamResource, but the StorageResource would be similar.
```
class MyStreamResource(ConfigurableResource):
    username: str
    password: str

    def get_batch(n=1000, ...):
        client = SomeClient(self.username, self.password)
        cursor = None
        while True:
             batch = client.get_some_records(after: cursor) 
             cursor = batch['cursor']
             yield batch['data']
```

and so on.

**Reply by AlejandroUPC** [reply]
Hey @PedramNavid 🐤 ,

I am simply persisting the messages in a database (not sure what you meant with How you solve this will depend on where you are persisting data, sorry)

That looks interesting and I agree that I have been struggling on switching my mentality from ops and jobs (tasks) to assets, will play around with this and let you know, thanks.

**Reply by PedramNavid** [reply]
Took me a while to wrap my head around it but once i did, it changed everything! make sure you check out launch week because we have two announcements that i think will help make the head wrapping easier 🦆

**Reply by vlreinier** [reply]
I am running into a similar scenario. I retrieve batches of files and dont want to store these in memory. However I see a few downsides to this approach. I do want to cache file processing steps to prevent rerunning expensive operations. I also really liked using partitioning as well to allow for progression status, fine grained control and execution in the web interface. What are my options? Do I still implement a single asset with my own caching? Or is there a better way to implement such a pipeline while keeping the benefits such as the retrying of partitions and monitoring the progress? TIA!

The pipeline looks like:

Step 1: generate batches of files
Step 2: process each batch and cache it to prevent processing it again
Step 3: Upload the processed files to a database.
